File: xls/dslx/fuzzer/ast_generator_test.py
# Lint as: python3
#
# Copyright 2020 The XLS Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Tests for xls.dslx.fuzzer.ast_generator."""

import random

from xls.common import test_base
from xls.dslx import ast_helpers
from xls.dslx import fakefs_test_util
from xls.dslx import parser_helpers
from xls.dslx.fuzzer import ast_generator
from xls.dslx.python import cpp_scanner as scanner
from xls.dslx.python import cpp_typecheck
from xls.dslx.python.import_routines import ImportCache
from xls.dslx.span import PositionalError


class AstGeneratorTest(test_base.TestCase):

  def test_generates_valid_functions(self):
    g = ast_generator.AstGenerator(
        random.Random(0), ast_generator.AstGeneratorOptions())
    for i in range(32):
      print('Generating sample', i)
      _, m = g.generate_function_in_module('main', 'test')
      text = str(m)
      filename = '/fake/test_sample.x'
      import_cache = ImportCache()
      additional_search_paths = ()
      with fakefs_test_util.scoped_fakefs(filename, text):
        try:
          module = parser_helpers.parse_text(
              text, name='test_sample', print_on_error=True, filename=filename)
          cpp_typecheck.check_module(module, import_cache,
                                     additional_search_paths)
        except PositionalError as e:
          parser_helpers.pprint_positional_error(e)
          raise

  def _test_repeatable(self, seed: int) -> None:

    def generate_one(g):
      _, m = g.generate_function_in_module('main', 'test')
      text = str(m)
      return text

    count_to_generate = 128
    g = ast_generator.AstGenerator(
        random.Random(seed), ast_generator.AstGeneratorOptions())
    fs = [generate_one(g) for _ in range(count_to_generate)]
    g.reset(seed=seed)
    gs = [generate_one(g) for _ in range(count_to_generate)]
    self.assertEqual(fs, gs)

    # Create a fresh generator as well.
    g = ast_generator.AstGenerator(
        random.Random(seed), ast_generator.AstGeneratorOptions())
    hs = [generate_one(g) for _ in range(count_to_generate)]
    self.assertEqual(gs, hs)

  def test_repeatable_seed0(self):
    self.maxDiff = None  # pylint: disable=invalid-name
    self._test_repeatable(seed=0)

  def test_repeatable_seed1(self):
    self.maxDiff = None  # pylint: disable=invalid-name
    self._test_repeatable(seed=1)

  def test_types(self):
    g = ast_generator.AstGenerator(
        random.Random(0), ast_generator.AstGeneratorOptions())
    _, m = g.generate_function_in_module('main', 'test')

    u1 = g._make_type_annotation(signed=False, width=1)
    self.assertEqual(g._get_type_bit_count(u1), 1)
    self.assertTrue(g._is_bit_vector(u1))
    self.assertTrue(g._is_unsigned_bit_vector(u1))
    self.assertFalse(g._is_array(u1))
    self.assertFalse(g._is_tuple(u1))

    s1 = g._make_type_annotation(signed=True, width=1)
    self.assertTrue(g._is_bit_vector(s1))
    self.assertFalse(g._is_unsigned_bit_vector(s1))
    self.assertEqual(g._get_type_bit_count(s1), 1)
    self.assertFalse(g._is_array(s1))
    self.assertFalse(g._is_tuple(s1))

    u42 = g._make_type_annotation(signed=False, width=42)
    self.assertEqual(g._get_type_bit_count(u42), 42)
    self.assertTrue(g._is_bit_vector(u42))
    self.assertTrue(g._is_unsigned_bit_vector(u42))
    self.assertFalse(g._is_array(u42))
    self.assertFalse(g._is_tuple(u42))

    empty_tuple = g._make_tuple_type(())
    self.assertEqual(g._get_type_bit_count(empty_tuple), 0)
    self.assertFalse(g._is_bit_vector(empty_tuple))
    self.assertFalse(g._is_unsigned_bit_vector(empty_tuple))
    self.assertFalse(g._is_array(empty_tuple))
    self.assertTrue(g._is_tuple(empty_tuple))

    tple = g._make_tuple_type((empty_tuple, s1, u1, u42, u42))
    self.assertEqual(g._get_type_bit_count(tple), 86)
    self.assertFalse(g._is_bit_vector(tple))
    self.assertFalse(g._is_unsigned_bit_vector(tple))
    self.assertFalse(g._is_array(tple))
    self.assertTrue(g._is_tuple(tple))

    nested_tuple = g._make_tuple_type((tple, u42, tple))
    self.assertEqual(g._get_type_bit_count(nested_tuple), 214)
    self.assertFalse(g._is_bit_vector(nested_tuple))
    self.assertFalse(g._is_unsigned_bit_vector(nested_tuple))
    self.assertFalse(g._is_array(nested_tuple))
    self.assertTrue(g._is_tuple(nested_tuple))

    array_of_u42 = g._make_array_type(u42, 100)
    self.assertEqual(g._get_type_bit_count(array_of_u42), 4200)
    self.assertFalse(g._is_bit_vector(array_of_u42))
    self.assertFalse(g._is_unsigned_bit_vector(array_of_u42))
    self.assertEqual(g._get_array_size(array_of_u42), 100)
    self.assertTrue(g._is_array(array_of_u42))
    self.assertFalse(g._is_tuple(array_of_u42))

    array_of_tuple = g._make_array_type(nested_tuple, 10)
    self.assertEqual(g._get_type_bit_count(array_of_tuple), 2140)
    self.assertFalse(g._is_bit_vector(array_of_tuple))
    self.assertFalse(g._is_unsigned_bit_vector(array_of_tuple))
    self.assertEqual(g._get_array_size(array_of_tuple), 10)
    self.assertTrue(g._is_array(array_of_tuple))
    self.assertFalse(g._is_tuple(array_of_tuple))

    u11_token = scanner.Token(g.fake_span, scanner.KeywordFromString('u11'))
    u11 = ast_helpers.make_builtin_type_annotation(
        m, g.fake_span, u11_token, dims=())
    self.assertEqual(g._get_type_bit_count(u11), 11)
    self.assertTrue(g._is_bit_vector(u11))
    self.assertTrue(g._is_unsigned_bit_vector(u11))
    self.assertFalse(g._is_array(u11))
    self.assertFalse(g._is_tuple(u11))

    un_token = scanner.Token(g.fake_span, scanner.KeywordFromString('uN'))
    un1234 = ast_helpers.make_builtin_type_annotation(
        m, g.fake_span, un_token, dims=(g._make_number(1234, None),))
    self.assertEqual(g._get_type_bit_count(un1234), 1234)
    self.assertTrue(g._is_bit_vector(un1234))
    self.assertTrue(g._is_unsigned_bit_vector(un1234))
    self.assertFalse(g._is_array(un1234))
    self.assertFalse(g._is_tuple(un1234))

    sn_token = scanner.Token(g.fake_span, scanner.KeywordFromString('sN'))
    sn1234 = ast_helpers.make_builtin_type_annotation(
        m, g.fake_span, sn_token, dims=(g._make_number(1234, None),))
    self.assertEqual(g._get_type_bit_count(sn1234), 1234)
    self.assertTrue(g._is_bit_vector(sn1234))
    self.assertFalse(g._is_unsigned_bit_vector(sn1234))
    self.assertFalse(g._is_array(sn1234))
    self.assertFalse(g._is_tuple(sn1234))

    bits_token = scanner.Token(g.fake_span, scanner.KeywordFromString('bits'))
    bits1234 = ast_helpers.make_builtin_type_annotation(
        m, g.fake_span, bits_token, dims=(g._make_number(1234, None),))
    self.assertEqual(g._get_type_bit_count(bits1234), 1234)
    self.assertTrue(g._is_bit_vector(bits1234))
    self.assertTrue(g._is_unsigned_bit_vector(bits1234))
    self.assertFalse(g._is_array(bits1234))
    self.assertFalse(g._is_tuple(bits1234))

    un1234_10 = ast_helpers.make_builtin_type_annotation(
        m,
        g.fake_span,
        un_token,
        dims=(g._make_number(1234, None), g._make_number(10, None)))
    self.assertEqual(g._get_type_bit_count(un1234_10), 12340)
    self.assertFalse(g._is_bit_vector(un1234_10))
    self.assertFalse(g._is_unsigned_bit_vector(un1234_10))
    self.assertTrue(g._is_array(un1234_10))
    self.assertFalse(g._is_tuple(un1234_10))


if __name__ == '__main__':
  test_base.main()

File: xls/fuzzer/ast_generator_test.cc
// Copyright 2021 The XLS Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "xls/fuzzer/ast_generator.h"

#include <cstdint>
#include <memory>
#include <optional>
#include <random>
#include <string>
#include <string_view>
#include <vector>

#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/log/check.h"
#include "absl/log/log.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/match.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/str_format.h"
#include "xls/common/logging/log_lines.h"
#include "xls/common/status/matchers.h"
#include "xls/common/status/status_macros.h"
#include "xls/dslx/command_line_utils.h"
#include "xls/dslx/create_import_data.h"
#include "xls/dslx/frontend/ast.h"
#include "xls/dslx/frontend/pos.h"
#include "xls/dslx/parse_and_typecheck.h"
#include "xls/dslx/virtualizable_file_system.h"
#include "re2/re2.h"

namespace xls::dslx {
namespace {

using ::testing::ContainsRegex;

// Parses and typechecks the given text to ensure it's valid -- prints errors to
// the screen in a useful way for debugging if they fail parsing / typechecking.
template <typename ModuleMember>
absl::Status ParseAndTypecheck(std::string_view text,
                               std::string_view module_name) {
  XLS_LOG_LINES(INFO, text);

  std::string filename = absl::StrCat(module_name, ".x");

  auto import_data = CreateImportDataForTest();
  absl::StatusOr<TypecheckedModule> parsed = ParseAndTypecheck(
      text, /*path=*/filename, /*module_name=*/module_name, &import_data);

  UniformContentFilesystem vfs(text, /*expect_path=*/filename);
  TryPrintError(parsed.status(), import_data.file_table(), vfs);
  XLS_RETURN_IF_ERROR(
      parsed->module->GetMemberOrError<ModuleMember>("main").status());
  return absl::OkStatus();
}

}  // namespace

class AstGeneratorTest : public ::testing::Test {
 protected:
  FileTable file_table_;
  AstGeneratorOptions options_;
  std::mt19937_64 rng_{0};
  AstGenerator g_{options_, rng_, file_table_};
};

TEST_F(AstGeneratorTest, BitsTypeGetMetadata) {
  g_.module_ = std::make_unique<Module>("test_module", /*fs_path=*/std::nullopt,
                                        file_table_);

  TypeAnnotation* u7 = g_.MakeTypeAnnotation(/*is_signed=*/false, 7,
                                             /*use_xn=*/false);
  LOG(INFO) << "u7: " << u7->ToString();
  XLS_ASSERT_OK_AND_ASSIGN(int64_t bit_count, g_.BitsTypeGetBitCount(u7));
  XLS_ASSERT_OK_AND_ASSIGN(bool is_signed, g_.BitsTypeIsSigned(u7));
  EXPECT_EQ(bit_count, 7);
  EXPECT_EQ(is_signed, false);

  TypeAnnotation* s129 = g_.MakeTypeAnnotation(/*is_signed=*/true, 129,
                                               /*use_xn=*/false);
  LOG(INFO) << "s129: " << s129->ToString();
  XLS_ASSERT_OK_AND_ASSIGN(bit_count, g_.BitsTypeGetBitCount(s129));
  XLS_ASSERT_OK_AND_ASSIGN(is_signed, g_.BitsTypeIsSigned(s129));
  EXPECT_EQ(bit_count, 129);
  EXPECT_EQ(is_signed, true);

  TypeAnnotation* xn_true_128 =
      g_.MakeTypeAnnotation(/*is_signed=*/true, 128, /*use_xn=*/true);
  EXPECT_EQ(xn_true_128->ToString(), "xN[bool:0x1][128]");
  XLS_ASSERT_OK_AND_ASSIGN(bit_count, g_.BitsTypeGetBitCount(xn_true_128));
  EXPECT_EQ(bit_count, 128);
  XLS_ASSERT_OK_AND_ASSIGN(is_signed, g_.BitsTypeIsSigned(xn_true_128));
  EXPECT_EQ(is_signed, true);
}

TEST_F(AstGeneratorTest, GeneratesParametricBindings) {
  g_.module_ =
      std::make_unique<Module>("my_mod", /*fs_path=*/std::nullopt, file_table_);
  std::vector<ParametricBinding*> pbs = g_.GenerateParametricBindings(4);
  EXPECT_EQ(pbs.size(), 4);
  // Note that the fact we get an unsigned binding is probabilistic, so we
  // generate four examples to try to find one that matches our unsigned
  // pattern.
  //
  // TODO(https://github.com/google/googletest/issues/3084): 2021-08-12
  // googletest cannot currently seem to use \d in regexp patterns, which is
  // quite surprising.
  constexpr const char* kWantPattern =
      R"(x[0-9]+: u[0-9]+ = \{u[0-9]+:0[xb][0-9a-f_]+\})";
  bool found_match = false;
  for (ParametricBinding* pb : pbs) {
    if (RE2::FullMatch(pb->ToString(), kWantPattern)) {
      found_match = true;
      break;
    }
  }
  EXPECT_TRUE(found_match);
}

namespace {
// Simply tests that we generate a bunch of valid functions using seed 0 (that
// parse and typecheck).
TEST(AstGeneratorMultiTest, GeneratesValidFunctions) {
  FileTable file_table;
  std::mt19937_64 rng{0};
  AstGeneratorOptions options;
  for (int64_t i = 0; i < 32; ++i) {
    AstGenerator g(options, rng, file_table);
    LOG(INFO) << "Generating sample: " << i;
    std::string module_name = absl::StrFormat("sample_%d", i);
    XLS_ASSERT_OK_AND_ASSIGN(AnnotatedModule module,
                             g.Generate("main", module_name));
    std::string text = module.module->ToString();
    // Parses/typechecks as well, which is primarily what we're testing here.
    XLS_ASSERT_OK(ParseAndTypecheck<Function>(text, module_name));
  }
}

// Simply tests that we generate a bunch of valid procs with an empty state type
// using seed 0 (that parse and typecheck).
TEST(AstGeneratorMultiTest, GeneratesValidProcsWithEmptyState) {
  FileTable file_table;
  std::mt19937_64 rng{0};
  AstGeneratorOptions options;
  options.generate_proc = true;
  options.emit_stateless_proc = true;
  // Regex matcher for the next function signature.
  constexpr const char* kWantPattern = R"(next\(x[0-9]+: \(\)\))";
  for (int64_t i = 0; i < 32; ++i) {
    AstGenerator g(options, rng, file_table);
    LOG(INFO) << "Generating sample: " << i;
    std::string module_name = absl::StrFormat("sample_%d", i);
    XLS_ASSERT_OK_AND_ASSIGN(AnnotatedModule module,
                             g.Generate("main", module_name));

    std::string text = module.module->ToString();
    //  Parses/typechecks as well, which is primarily what we're testing here.
    XLS_ASSERT_OK(ParseAndTypecheck<Proc>(text, module_name)) << text;
    EXPECT_THAT(text, ContainsRegex(kWantPattern));
  }
}

// Simply tests that we generate a bunch of valid procs with a random state type
// using seed 0 (that parse and typecheck).
TEST(AstGeneratorMultiTest, GeneratesValidProcsWithRandomState) {
  FileTable file_table;
  std::mt19937_64 rng{0};
  AstGeneratorOptions options;
  options.generate_proc = true;
  // Regex matcher for the next function signature.
  // Although [[:word:]] encapsulates [0-9a-zA-Z_], which would simplify the
  // following regex statement. The following regex statement is more readable.
  constexpr const char* kWantPattern =
      R"(next\(x[0-9]+: []0-9a-zA-Z_, ()[]+\))";
  for (int64_t i = 0; i < 32; ++i) {
    AstGenerator g(options, rng, file_table);
    LOG(INFO) << "Generating sample: " << i;
    std::string module_name = absl::StrFormat("sample_%d", i);
    XLS_ASSERT_OK_AND_ASSIGN(AnnotatedModule module,
                             g.Generate("main", module_name));

    std::string text = module.module->ToString();
    //  Parses/typechecks as well, which is primarily what we're testing here.
    XLS_ASSERT_OK(ParseAndTypecheck<Proc>(text, module_name));
    EXPECT_THAT(text, ContainsRegex(kWantPattern));
  }
}

// Helper function that is used in a TEST_P so we can shard the work.
static void TestRepeatable(uint64_t seed) {
  FileTable file_table;
  AstGeneratorOptions options;
  // Capture first output at a given seed for comparison.
  std::optional<std::string> first;
  // Try 32 generations at a given seed.
  for (int64_t i = 0; i < 32; ++i) {
    std::mt19937_64 rng{seed};
    AstGenerator g(options, rng, file_table);
    XLS_ASSERT_OK_AND_ASSIGN(AnnotatedModule module,
                             g.Generate("main", "test"));
    std::string text = module.module->ToString();
    if (first.has_value()) {
      ASSERT_EQ(text, *first) << "sample " << i << " seed " << seed;
    } else {
      first = text;
      // Parse and typecheck for good measure.
      XLS_ASSERT_OK(ParseAndTypecheck<Function>(text, "test"));
    }
  }
}

TEST(AstGeneratorMultiTest, GeneratesZeroWidthValues) {
  FileTable file_table;
  std::mt19937_64 rng{0};
  AstGeneratorOptions options;
  options.emit_zero_width_bits_types = true;
  bool saw_zero_width = false;
  // Every couple samples seems to produce a zero-width value somewhere, but set
  // to a high number to catch invalid handling of zero-width values in the
  // generator.
  constexpr int64_t kNumSamples = 5000;
  for (int64_t i = 0; i < kNumSamples; ++i) {
    AstGenerator g(options, rng, file_table);
    VLOG(1) << "Generating sample: " << i;
    std::string module_name = absl::StrFormat("sample_%d", i);
    XLS_ASSERT_OK_AND_ASSIGN(AnnotatedModule module,
                             g.Generate("main", module_name));
    std::string text = module.module->ToString();
    if (absl::StrContains(text, "uN[0]") || absl::StrContains(text, "sN[0]")) {
      VLOG(1) << absl::StrFormat("Saw zero-width type after %d samples", i);
      saw_zero_width = true;
    }
  }
  EXPECT_TRUE(saw_zero_width) << absl::StrFormat(
      "Generated %d samples and did not see a zero-width type", kNumSamples);
}

class AstGeneratorRepeatableTest : public testing::TestWithParam<uint64_t> {};

TEST_P(AstGeneratorRepeatableTest, GenerationRepeatableAtSeed) {
  TestRepeatable(/*seed=*/GetParam());
}

INSTANTIATE_TEST_SUITE_P(AstGeneratorRepeatableTestInstance,
                         AstGeneratorRepeatableTest,
                         testing::Range(uint64_t{0}, uint64_t{1024}));

}  // namespace

}  // namespace xls::dslx
